### ISSUE TITLE
storage: show-connection-string incorrect for Azure Stack storage accounts

### ISSUE THREAD

**Issue by akshayjshah** [Azure/azure-cli#18413]
## Describe the bug

**Command Name**
`az storage account show-connection-string`

**Errors:**
The `EndpointSuffix` in the connection string returned by the CLI is wrong: the storage account I'm querying is deployed in an Azure Stack product and should have an `EndpointSuffix` ending in `azure.net`, but I'm getting the standard suffix for public Azure (`core.windows.net`).

```sh
$ az storage account show-connection-string -n mystackstorage | jq .
{
  "connectionString": "DefaultEndpointsProtocol=https;EndpointSuffix=core.windows.net;AccountName=mystackstorage;AccountKey=<redacted>"
}
```

## To Reproduce:
Steps to reproduce the behavior. Note that argument values have been redacted, as they may contain sensitive information.

- Create a storage account called `mystackstorage` in Azure Stack (see Additional Context, below)
- `az storage account show-connection-string -n mystackstorage`

## Expected Behavior

I expected to see the same `EndpointSuffix` that I'd get by copy-pasting from the Keys section of the Azure Portal.

I traced the problem as far as https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/storage/operations/account.py#L224, where we set `endpoint_suffix` to the cloud's standard storage endpoint suffix. In my case, I'm using the Azure public cloud, so this is always the static value set in https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/cloud.py#L320. It looks to me like we should be getting this information from a network request instead.

## Environment Summary
```
Linux-5.4.72-microsoft-standard-WSL2-x86_64-with-debian-bullseye-sid
Python 3.6.10
Installer: DEB

azure-cli 2.24.1 *

Extensions:
account 0.2.1

```
## Additional Context

I work at Microsoft on the Azure Stack family of products. Some of the details required to reproduce this bug are specific to a project that isn't public yet - happy to discuss more details internally.

/cc @adepue @DwayneNeed @shriramnat @ChrisDickensMSFT

**Comment by yonzhan** [comment]
storage

**Comment by Juliehzl** [comment]
Hi @akshayjshah, you need to configure your cloud first. Azure CLI only provides public and gov cloud information. To configure your cloud,  there are two ways:
- `az cloud register` to register your Azure Stack could  and add cloud endpoints information
- If you have a url with cloud information, you could set env variable ARM_CLOUD_METADATA_URL and we would also retrieve cloud information from the target url.

**Comment by akshayjshah** [comment]
I don't think that should be necessary: I'm deploying to [Azure Stack Edge Zones](https://azure.microsoft.com/en-us/blog/microsoft-partners-with-the-industry-to-unlock-new-5g-scenarios-with-azure-edge-zones/), not Azure Stack Hub. I'm deploying resources to the edge zone using resource providers from the standard Azure public cloud, and the `az group`, `az deployment`, `az vm`, `az cosmosdb`, etc. commands can all interact with resources deployed in the Edge Zone using the public cloud configuration.

**Comment by Juliehzl** [comment]
> I don't think that should be necessary: I'm deploying to [Azure Stack Edge Zones](https://azure.microsoft.com/en-us/blog/microsoft-partners-with-the-industry-to-unlock-new-5g-scenarios-with-azure-edge-zones/), not Azure Stack Hub. I'm deploying resources to the edge zone using resource providers from the Azure public cloud, and the `az group`, `az deployment`, `az vm`, `az cosmosdb`, etc. commands can all interact with resources deployed in the Edge Zone using the public cloud configuration.

I see. For Azure Stack Edge Zones related settings, currently you have to configure endpoints with `az cloud update` for right edge zone information in storage, network , resource and etc. For getting right connection endpoints, currently Azure CLI doesn't have the capability to get edge zone endpoint automatically and users have to config by themselves. 
But we will add the capability when there is available API for edge zone information and try to learn from Portal.

**Comment by akshayjshah** [comment]
> But we will add the capability when there is available API for edge zone information and try to learn from Portal.

@DwayneNeed Do you know if there's already a Portal API that exposes the right information?

**Comment by DwayneNeed** [comment]
This thread is confusing because this is actually about **edge zones** not Azure Stack.  Can someone correct the title?

For a specific storage account, you can get the properties and determine the endpoint DNS name:

https://docs.microsoft.com/en-us/rest/api/storagerp/storage-accounts/get-properties

```
{
  "extendedLocation": {
    "type": "EdgeZone",
    "name": "microsoftlosangeles1"
  },
  "sku": {
    "name": "Premium_LRS",
    "tier": "Premium"
  },
  "kind": "BlockBlobStorage",
  "id": "/subscriptions/baffff89-2c76-424a-af4c-43b7518a3cb8/resourceGroups/akshay-edge-demo-01/providers/Microsoft.Storage/storageAccounts/akshay74d1",
  "name": "akshay74d1",
  "type": "Microsoft.Storage/storageAccounts",
  "location": "westus",
  "tags": {},
  "properties": {
    "keyCreationTime": {
      "key1": "2021-07-19T21:51:46.6137965Z",
      "key2": "2021-07-19T21:51:46.6137965Z"
    },
    "primaryExtendedLocation": {
      "type": "EdgeZone",
      "name": "microsoftlosangeles1"
    },
    "privateEndpointConnections": [],
    "allowBlobPublicAccess": true,
    "allowSharedKeyAccess": true,
    "networkAcls": {
      "bypass": "AzureServices",
      "virtualNetworkRules": [],
      "ipRules": [],
      "defaultAction": "Allow"
    },
    "supportsHttpsTrafficOnly": true,
    "encryption": {
      "services": {
        "file": {
          "keyType": "Account",
          "enabled": true,
          "lastEnabledTime": "2021-07-19T21:51:46.6137965Z"
        },
        "blob": {
          "keyType": "Account",
          "enabled": true,
          "lastEnabledTime": "2021-07-19T21:51:46.6137965Z"
        }
      },
      "keySource": "Microsoft.Storage"
    },
    "accessTier": "Hot",
    "provisioningState": "Succeeded",
    "creationTime": "2021-07-19T21:51:46.5043498Z",
    "primaryEndpoints": {
      "dfs": https://akshay74d1.dfs.microsoftlosangeles1.edgestorage.azure.net/,
      "web": https://akshay74d1.web.microsoftlosangeles1.edgestorage.azure.net/,
      "blob": https://akshay74d1.blob.microsoftlosangeles1.edgestorage.azure.net/
    },
    "primaryLocation": "westus",
    "statusOfPrimary": "available"
  }
}
```

Note the blob URL:
**"blob": https://akshay74d1.blob.microsoftlosangeles1.edgestorage.azure.net/**

**Comment by akshayjshah** [comment]
> This thread is confusing because this is actually about edge zones not Azure Stack. Can someone correct the title?

Sorry, Dwayne - updated the title and the top-level description of the problem.
